### PR TITLE
derive R.take and R.drop from R.slice

### LINKED ISSUE #1158: derive R.take and R.drop from R.slice
`R.slice` works on strings as well as arrays, since we dispatch to a `slice` method if present:

```javascript
> R.slice(0, 3, [1, 2, 3, 4, 5])
[ 1, 2, 3 ]
> R.slice(0, 3, 'Ramda')
'Ram'
```

One can define `take` in terms of `slice`. If one does so `take` also works on strings:

```javascript
> var take = R.slice(0)
undefined
> take(3, [1, 2, 3, 4, 5])
[ 1, 2, 3 ]
> take(3, 'Ramda')
'Ram'
```

`drop` can be defined similarly:

```javascript
> var drop = R.slice(R.__, Infinity)
undefined
> drop(3, [1, 2, 3, 4, 5])
[ 4, 5 ]
> drop(3, 'Ramda')
'da'
```

`take` and `drop` are derivable from `slice`. I suggest we update the implementations of `R.take` and `R.drop` to use this approach. We could then deprecate `R.substringTo` and `R.substringFrom`.

### PR DESCRIPTION
**Pull request by davidchambers**
Closes #1158

The simplicity of the implementation reinforces my belief that this is an entirely reasonable change. :)

### CODE REVIEW

File: src/drop.js [hunk 1]
```diff
@@ -14,13 +14,13 @@ var _xdrop = require('./internal/_xdrop');
  * @memberOf R
  * @category List
  * @sig Number -> [a] -> [a]
- * @param {Number} n The number of elements of `list` to skip.
- * @param {Array} list The array to consider.
- * @return {Array} The last `n` elements of `list`.
```
Comment (davidchambers) [inline src/drop.js]: This was erroneous.

File: src/drop.js [hunk 2]
```diff
@@ -14,13 +14,13 @@ var _xdrop = require('./internal/_xdrop');
...
+ * @param {Array} xs The collection to consider.
+ * @return {Array}
  * @example
  *
  *      R.drop(3, [1,2,3,4,5,6,7]); //=> [4,5,6,7]
  */
-module.exports = _curry2(_dispatchable('drop', _xdrop, function drop(n, list) {
-  return n <= 0 ? list : _slice(list, n);
+module.exports = _curry2(_dispatchable('drop', _xdrop, function drop(n, xs) {
+  return n <= 0 ? xs : slice(n, Infinity, xs);
```
Comment (buzzdecafe) [inline src/drop.js]: didn't we agree somewhere always to clone? in which case `n < 0` might be good enough

File: src/take.js [hunk 3]
```diff
@@ -28,6 +28,6 @@ var _xtake = require('./internal/_xtake');
  *      var takeFive = R.take(5);
  *      takeFive(members); //=> ["Paul Desmond","Bob Bates","Joe Dodge","Ron Crotty","Lloyd Davis"]
  */
-module.exports = _curry2(_dispatchable('take', _xtake, function take(n, list) {
-  return _slice(list, 0, n);
+module.exports = _curry2(_dispatchable('take', _xtake, function take(n, xs) {
+  return n < 0 ? xs : slice(0, n, xs);
```
Comment (buzzdecafe) [inline src/take.js]: (likewise here)

File: test/drop.js [hunk 4]
```diff
@@ -18,6 +18,13 @@ describe('drop', function() {
     assert.deepEqual(R.drop(-Infinity, [1, 2, 3]), [1, 2, 3]);
   });
 
+  it('dispatches to `slice` method', function() {
+    var o = {slice: function(a, b) { return '[' + a + ':' + b + ']'; }};
+
+    assert.strictEqual(R.drop(3, 'Ramda'), 'da');
+    assert.strictEqual(R.drop(3, o), '[3:Infinity]');
+  });
+
```
Comment (buzzdecafe) [inline test/drop.js]: to satisfy https://github.com/ramda/ramda/issues/1089 should probably have a test of that sort

**Comment by buzzdecafe** [comment]
this is simple in implementation, and that is nice. and reducing the footprint of the api is also a plus. as noted in #1158 i think we should tread carefully when deliberately treating strings as lists. then shouldn't all list functions work on strings? i certainly wouldn't want to support that. So I'm a bit torn on this. Curious to hear other opinions.

File: src/drop.js (hunk 2 above)
Comment (davidchambers) [inline src/drop.js]: :zap:

File: src/take.js (hunk 3 above)
Comment (davidchambers) [inline src/take.js]: :zap:

File: test/drop.js (hunk 4 above)
Comment (davidchambers) [inline test/drop.js]: :zap:

**Comment by megawac** [comment]
> when deliberately treating strings as lists

Could be done pretty easily with a wrapper I agree, its a double edge sword. On one hand, I would love to be able to treat strings as list of chars without thinking of it; on the other hand its significant complexity

**Comment by CrossEye** [comment]
I agree that we don't want to make the wholesale transition to Strings as lists, but I see much less harm in doing that where we can than in the notion some libraries have of treating Objects in the same way.  Many of the list functions make sense when applied to Strings.  In fact, I think pretty much any ` ? -> [a] -> ?` would comfortably apply to Strings.  I just don't see any great reason to go out of our way to do this.  When it can fall out of what we're doing, I have no issue, and this -- building two of our functions on top of an existing internal function, thereby removing the need for two of our less intuitive ones -- seems like a no-brainer.

So a :herb: from me, with the understanding that I agree with @buzzdecafe about not making this a precedent for turning all possible list functions into functions on Strings as well.

**Comment by buzzdecafe** [comment]
thanks, :cat: :cat:. let's keep an eye out for string-creep. but in the meantime :cow2:

**Comment by davidchambers** [comment]
> then shouldn't all list functions work on strings?

Due to dynamic dispatch several Ramda functions already work on strings: `R.concat`, `R.indexOf`, and `R.slice` come to mind. If one day the language defines `String.prototype.contains`, I'll expect `R.contains` to work on strings too.

I agree with you both, though, that we should not have any string-specific logic in our list functions. This is a nice clear line to draw.

**Comment by megawac** [comment]
FYI, String.prototype.includes exists in the ES6 spec

**Comment by davidchambers** [comment]
> FYI, String.prototype.includes exists in the ES6 spec

Thanks, @megawac. I've just realized that `contains` is derivable from `indexOf`, so `R.contains` *can* work on strings if we define it appropriately.

**Comment by paldepind** [comment]
I'm happy to see `substringTo` and `substringFrom` deprecated. But I'm a little unhappy too see that this deliberately adds supports for strings to a function without it being documented. The type signature still says `Array` and it is completely opaque that this function will work on strings because they have a `slice` method.

**Comment by davidchambers** [comment]
The lack of documentation was an oversight on my part. I'll rectify the situation with another pull request this weekend.

**Comment by paldepind** [comment]
Great! No problem :)

**Comment by asaf-romano** [comment]
Hrm, am I right that the string-based API supported negative offset, but drop doesn't?

**Comment by davidchambers** [comment]
`R.slice` accepts negative offsets:

```javascript
> R.slice(-3, Infinity, 'foobar')
'bar'
```

**Pull request merged**